tests/integration/mounts.bats

```diff
@@ -96,14 +108,101 @@ function teardown() {
 	test_ro_cgroup_mount
 }
 
-# https://github.com/opencontainers/runc/security/advisories/GHSA-m8cg-xc2p-r3fc
-# shellcheck disable=SC2031
-function test_ro_cgroup_mount() {
-	local lines status
-	# shellcheck disable=SC2016
-	update_config '.process.args |= ["sh", "-euc", "for f in `grep /sys/fs/cgroup /proc/mounts | awk \"{print \\\\$2}\"| uniq`; do test -e $f && grep -w $f /proc/mounts | tail -n1; done"]'
+function config_add_bind_mount() {
+	src="$1"
+	dst="$2"
+	bindtype="${3:-bind}"
+	idmap="${4:-}"
+
+	mappings=""
+	if [ -n "$idmap" ]; then
+		mappings='
+			"uidMappings": [{"containerID": 0, "hostID": 100000, "size": 65536}],
+			"gidMappings": [{"containerID": 0, "hostID": 100000, "size": 65536}],
+		'
+	fi
+
+	update_config '.mounts += [{
+		"source": "'"$src"'",
+		"destination": "'"$dst"'",
+		"type": "bind",
+		'"$mappings"'
+		"options": [ "'"$bindtype"'", "rprivate" ]
+	}]'
+}
+
+function test_mount_order() {
+	for flag in userns idmap; do
+		is_set=
+		if grep -q "$flag" <<<"${1:-}"; then
+			is_set=1
+		fi
+		eval "is_$flag=$is_set"
+	done
+
+	from_ctr_bind_args=("rbind")
+
+	if [ -n "$is_userns" ]; then
+		requires root
+
+		update_config '.linux.namespaces += [{"type": "user"}]
+			| .linux.uidMappings += [{"containerID": 0, "hostID": 100000, "size": 65536}]
+			| .linux.gidMappings += [{"containerID": 0, "hostID": 100000, "size": 65536}]'
+		remap_rootfs
+	fi
+
+	if [ -n "$is_idmap" ]; then
+		requires root
+		requires_kernel 5.12
+		requires_idmap_fs .
+
+		from_ctr_bind_args+=("idmap")
+	fi
+
+	mkdir -p rootfs/{mnt,final}
+	# Create a set of directories we can create a mount tree with.
+	for subdir in a/x b/y c/z; do
+		dir="bind-src/$subdir"
+		mkdir -p "$dir"
+		echo "$subdir" > "$dir/file"
+	done
+
+	rootfs="$(pwd)/rootfs"
+	rm -rf rootfs/mnt
+	mkdir rootfs/mnt
+
+	# Create a bind-mount tree.
+	config_add_bind_mount "$PWD/bind-src/a" "/mnt"
+	config_add_bind_mount "$PWD/bind-src/b" "/mnt/x"
+	config_add_bind_mount "$PWD/bind-src/c" "/mnt/x/y"
+	config_add_bind_mount "$PWD/bind-src/a" "/mnt/x/y/z"
+	# Create a recursive bind-mount that uses part of the current tree in the
+	# container.
+	config_add_bind_mount "$rootfs/mnt/x" "$rootfs/mnt/x/y/z/x" "${from_ctr_bind_args[@]}"
+	config_add_bind_mount "$rootfs/mnt/x/y" "$rootfs/mnt/x/y/z" "${from_ctr_bind_args[@]}"
```

Comment:
What do we do in runc 1.1 with this bind-mount (I mean a bind-mount with a source inside the container rootfs)? Is the result exactly the same as now with this PR applied?

Reply:
runc 1.1 includes 9c444070ec7b, so with user namespaces it has the same error as current `main`.

The first test (`runc run [mount order, container bind-mount source]`) indicates the correct behaviour (we don't do anything special for bind-mounts when not in a user namespace) and thus the key thing is whether we have the same behaviour in the cases where we need to get mountfds. These tests pass on this PR, and don't pass on `1.1.10` or `main` because of 9c444070ec7b.

Reply:
Sure, for the bind mount sources case, but what for the other case (!bind mount source fds passing)? Are we keeping the same behavior as runc 1.1?

Reply:
Perfect, that addresses my concern. Thanks!